synapse/storage/event_push_actions.py

```diff
@@ -295,15 +296,17 @@ def get_unread_push_actions_for_user_in_range_for_email(
                 stream ordering of event push actions to fetch.
             max_stream_ordering(int): The inclusive upper bound on the
                 stream ordering of event push actions to fetch.
-            limit (int): The maximum number of rows to return.
+            msglimit (int): The maximum number of message push actions to return.
+            invlimit (int): The maximum number of invite push actions to return.
         Returns:
             A promise which resolves to a list of dicts with the keys "event_id",
             "room_id", "stream_ordering", "actions", "received_ts".
             The list will be ordered by descending received_ts.
-            The list will have between 0~limit entries.
+            The list will have between 0~(msglimit+invlimit) entries.
         """
         # find rooms that have a read receipt in them and return the most recent
-        # push actions
+        # as well as the oldest push actions (otherwise we might never send notifications
```

Comment:
I'm afraid I don't really follow what the bug you're trying to fix is, and this seems like an odd way to go about it.

Please could you try to explain the bug more clearly, and how this fixes it?

It's not obvious to me that I want a mix of old and new events in the email: it sounds confusing. Furthermore, doubling the number of database queries sounds unsatisfactory.

Reply:
Let's imagine a setting that sends mail after 10 minutes of no read receipt. The db query returns the most recent X events. If none of the X events are older than 10 minutes, no email will be sent. So, if we have a room in which there is constant activity so that the most recent X events are always less than 10 minutes old, no email will ever be sent. This is problematic especially when you set the mail delay to longer periods like 24h.